enable fp16 mkldnn fusion/prepack in inductor

zhuhaozhe opened the pull request:
- Extend `linear/conv/rnn` packable with `float16`.
- Extend `Unary fusion` to support `float16`.

Test Case:
    Extend bfloat16 related test in `test_cpu_repro.py` and `test_mkldnn_pattern_matcher.py` to test both `fp16` and `bf16`.



Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117206



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

zhuhaozhe committed:
enable fp16 mkldnn fusion/prepack in inductor

[ghstack-poisoned]

jgong5 commented on torch/_inductor/fx_passes/mkldnn_fusion.py:
```diff
@@ -68,13 +68,26 @@ def _to_bf16(input_call):
+        # only insert to_dtype if lp_dtype is True
         computation_call = (
-            _to_float(call_fn(), users=users) if is_bf16 else call_fn(users=users)
+            _to_float(call_fn(), users=users) if lp_dtype else call_fn(users=users)
```
Perhaps `lowp_dtype` is more readable?

zhuhaozhe replied:
Thanks for advice, changed `lp` to `lowp`

zhuhaozhe committed:
Update on "enable fp16 mkldnn fusion/prepack in inductor"

- Extend `linear/conv/rnn` packable with `float16`.
- Extend `Unary fusion` to support `float16`.

Test Case:
    Extend bfloat16 related test in `test_cpu_repro.py` and `test_mkldnn_pattern_matcher.py` to test both `fp16` and `bf16`.






cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

zhuhaozhe commented:
@pytorchbot merge

zhuhaozhe commented:
@pytorchbot merge
